markovflow/models/variational_cvi_sde.py

```diff
@@ -0,0 +1,348 @@
+
+from markovflow.models import MarkovFlowModel
+from markovflow.kalman_filter import UnivariateGaussianSitesNat
+from markovflow.sde.sde import SDE
+from markovflow.state_space_model import StateSpaceModel
+from markovflow.sde.sde_utils import linearize_sde
+from markovflow.emission_model import EmissionModel
+from markovflow.kalman_filter import KalmanFilterWithSparseSites
+from markovflow.sde.sde_utils import KL_sde
+from markovflow.models.variational_cvi import gradient_transformation_mean_var_to_expectation
+
+
+class CVISDESparseSites(MarkovFlowModel):
+    """
+    Provides a site-based parameterization to the variational posterior of a dynamical system.
+    """
+
+    def __init__(
+            self,
+            prior_sde: SDE,
+            grid: TensorType,
+            input_data: Tuple[TensorType, TensorType],
+            likelihood: Likelihood,
+            learning_rate=0.1
+    ) -> None:
+        """
+        :param prior_sde: Prior SDE over the latent states, x.
+        :param grid: Grid over time with shape ``batch_shape + [grid_size]``
+        :param input_data: A tuple containing the observed data:
+
+            * Time points of observations with shape ``batch_shape + [num_data]``
+            * Observations with shape ``batch_shape + [num_data, observation_dim]``
+
+        :param likelihood: A likelihood with shape ``batch_shape + [num_inducing]``.
+        :param learning_rate: The learning rate of the algorithm.
+        :param test_data: Test data used to calculate NLPD if not None.
+        """
+        super().__init__()
+
+        self._likelihood = likelihood
+        self.grid = grid
+        self.prior_sde = prior_sde
+
+        self._observations_time_points = input_data[0]
+        self._observations = input_data[1]
+
+        self.output_dim = 1
+        self.state_dim = self._observations.shape[-1]
+
+        self.sites = UnivariateGaussianSitesNat(
+            nat1=Parameter(tf.zeros_like(self._observations_time_points)[..., None]),
+            nat2=Parameter(tf.ones_like(self._observations_time_points)[..., None, None] * -1e-10),
+            log_norm=Parameter(tf.zeros_like(self._observations_time_points)[..., None]),
+        )
+        self.sites_lr = learning_rate
+
+        self._initialize_mean_statistic()
+        self.linearization_pnts = (tf.identity(self.fx_mus[:, :-1, :]), tf.identity(self.fx_covs[:, :-1, :, :]))
+        self._linearize_prior()
+
+        self.obs_sites_indices = tf.where(tf.equal(self.grid[..., None], self._observations_time_points))[:, 0][
+            ..., None]
+        self.dt = self.grid[1] - self.grid[0]
+
+    def _initialize_mean_statistic(self):
+        """
+        Initialize the prior on the initial state and the posterior statistics.
+        """
+        self.initial_mean = tf.zeros((1, self.state_dim), dtype=self._observations.dtype)
+        self.initial_chol_cov = tf.linalg.cholesky(tf.reshape(self.prior_sde.q, (1, self.state_dim, self.state_dim)))
```

Comment:
what is the initial_chol_cov set to the prior diffusion?
It would be better to have a setter,
and potentially pass the arguments to the model init.

Reply:
So, this is where I was a bit confused. In this function, we initialize the prior on the initial state by `self.initial_mean, self.initial_chol_cov` and the initial posterior `self.fx_mus, self.fx_covs` that is used to linearize the prior at the start. 

How should they be initialized? Like passed in the init with the default value being?

Reply:
I added options for it in the latest commit!